Strange behavior with comptime reflection

I did not expected to see this calls produced with comptime reflection to be inferred as `string`:
```v
string x = main__Foo_a(f ); // vweb action call with args;
```

Am i missing something or is this some sort of a bug?

Thanks

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version: V 0.2.4 12ffe04** 
**OS: linux, Ubuntu 21.10 (VM)** 

**What did you do?**
```v
struct Foo { }

fn (f Foo) a() int { 
    return 1
}

fn (f Foo) b() int {
    return 2
}

fn main() {
    f := Foo{}
    $for method in Foo.methods {
       x := f.$method()
    }
}
```

**What did you expect to see?**
The compilation succed
 
**What did you see instead?**
```
vagrant@ubuntu-impish:/vagrant$ v -keepc -cg issue.v
issue.v:14:8: warning: unused variable: `x`
   12 |     f := Foo{}
   13 |     $for method in Foo.methods {
   14 |        x := f.$method()
      |        ^
   15 |     }
   16 | }
/tmp/v_1000/issue.tmp.c:10703: error: '{' expected (got ";")
builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

```c
int main__Foo_a(main__Foo f) {
	int _t1 = 1;
	return _t1;
}

int main__Foo_b(main__Foo f) {
	int _t1 = 2;
	return _t1;
}

void main__main(void) {
	main__Foo f = (main__Foo){EMPTY_STRUCT_INITIALIZATION};
	/* $for method in main.Foo(methods) */ {
		FunctionData method = {0};
		/* method 0 */ {
			method.name = _SLIT("a");
			method.attrs = __new_array_with_default(0, 0, sizeof(string), 0);
			method.args = __new_array_with_default(0, 0, sizeof(MethodArgs), 0);
			method.typ = 0;
			method.return_type = 7;
			string x = main__Foo_a(f ); // vweb action call with args;
		}
		/* method 1 */ {
			method.name = _SLIT("b");
			method.attrs = __new_array_with_default(0, 0, sizeof(string), 0);
			method.args = __new_array_with_default(0, 0, sizeof(MethodArgs), 0);
			method.typ = 0;
			method.return_type = 7;
			string x = main__Foo_b(f ); // vweb action call with args;
		}
	}// $for
}
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3003,6 +3003,10 @@ fn (mut g Gen) gen_assign_stmt(assign_stmt ast.AssignStmt) {
 							left.obj.typ = var_type
 						}
 					}
+				} else if val is ast.ComptimeCall {
+					key_str := '${val.method_name}.return_type'
+					var_type = g.comptime_var_type_map[key_str] or { var_type }
+					left.obj.typ = var_type
 				}
 				is_auto_heap = left.obj.is_auto_heap
 			}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1146,12 +1146,12 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 	// g.generate_tmp_autofree_arg_vars(node, name)
 	// Handle `print(x)`
 	mut print_auto_str := false
-	if is_print && node.args[0].typ != ast.string_type { // && !free_tmp_arg_vars {
+	if is_print && (node.args[0].typ != ast.string_type || g.comptime_for_method.len > 0) { // && !free_tmp_arg_vars {
 		mut typ := node.args[0].typ
 		if typ == 0 {
 			g.checker_bug('print arg.typ is 0', node.pos)
 		}
-		if typ != ast.string_type {
+		if typ != ast.string_type || g.comptime_for_method.len > 0 {
 			expr := node.args[0].expr
 			typ_sym := g.table.get_type_symbol(typ)
 			if typ_sym.kind == .interface_ && (typ_sym.info as ast.Interface).defines_method('str') {

--- a/vlib/v/tests/comptime_for_method_call_test.v
+++ b/vlib/v/tests/comptime_for_method_call_test.v
@@ -1,0 +1,23 @@
+struct Foo {}
+
+fn (f Foo) a() int {
+	return 1
+}
+
+fn (f Foo) b() int {
+	return 2
+}
+
+fn test_comptime_for_method_call() {
+	f := Foo{}
+	mut rets := []string{}
+
+	$for method in Foo.methods {
+		x := f.$method()
+		println(x)
+		rets << x.str()
+	}
+	assert rets.len == 2
+	assert rets[0] == '1'
+	assert rets[1] == '2'
+}

